### PR TITLE
Improve unit handling in gammapy.catalog

### PR DESCRIPTION
**Pull request by cdeil**
This pull request:
- [x] Introduces `gammapy.utils.units` and `gammapy.utils.table` with helper functions to standardise units (get rid of "photon" and "counts").
- [x] Cleans up unit handling in gamma.catalog (reading units from the FITS file where given, instead of dropping it on read and then adding units back later via code)

### CODE REVIEW

File: gammapy/utils/units.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""Units and Quantity related helper functions"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+import astropy.units as u
+
+__all__ = [
+    'standardise_unit',
```
Comment (adrn) [inline gammapy/utils/units.py]: @mhvk may disagree, but I was just suggesting using the equivalency as a way to "standardise the unit" when you read, without having to go digging through the bases. For example:

```python
q1 = 10 * u.photon / u.cm**2 / u.s
q2 = 10 * u.count / u.cm**2 / u.s

def standardise_unit(q):
    eqs = [(u.Unit("ph/(cm2 s)"), u.Unit("1/(cm2 s)")),
           (u.Unit("ct/(cm2 s)"), u.Unit("1/(cm2 s)"))]
    return q.to("cm-2 s-1", equivalencies=eqs)

standardise_unit(q1)
```

Comment (mhvk) [inline gammapy/utils/units.py]: @cdeil - I think it depends a little on exactly how you want to apply it. If you know what output unit you want, I think @adrn's suggestion is best, as it is nice and explicit. If you simply want to remove all occurrences of `ph` and `ct`, then I think your approach is better.

**Comment by mhvk** [comment]
@cdeil - inline, I just answered how best to do what you wanted, but two comments more generally:
- Do use `QTable` -- it will use `Quantity` instead of `Column`, which means units will actually be properly used. I think this will save you much pain in the future...
- About this PR: don't the `photon` and `count` actually have different physical meaning? From both my Chandra/XMM days as well as optical work, I got to think of photon as having the effective area removed (possibly even the whole response matrix), while count has not. By keeping that information around, you can still distinguish the two. Might it not be better to force being explicit about this difference? (This is one part where @adrn's suggestion of an equivalency with dimensionless could come in handy: allow one of the two to be converted, but raise an error on the other.)

File: gammapy/utils/units.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    >>> from gammapy.utils.units import standardise_unit
+    >>> standardise_unit('ph cm-2 s-1')
+    Unit("1 / (cm2 s)")
+    >>> standardise_unit('ct cm-2 s-1')
+    Unit("1 / (cm2 s)")
+    >>> standardise_unit('cm-2 s-1')
+    Unit("1 / (cm2 s)")
+    """
+    unit = u.Unit(unit)
+    for base, power in zip(unit.bases, unit.powers):
```
Comment (mhvk) [inline gammapy/utils/units.py]: If you go this route, I would actually create new bases and powers and just make a new `CompositeUnit` with those (with the same scale). Though what you have of course essentially does the same thing under the hood.

**Comment by cdeil** [comment]
@adrn @mhvk - Thanks for the feedback!

A mix of "photon" and "counts" is no good in my experience (often we don't know which counts are photons or background).
I am tempted to standardise to always use "photon" in Gammapy, instead of dropping it.
The problem with that approach is that it's not possible to add "photon" in the right places in a generic automatic way on read, like it is to drop it as we do here.
So I think we would end up with options when and where to add "photon" in many places.

So for now, I'm pursuing the approach to drop it, and following the suggestion of @mhvk I've changed the implementation to this:
```python
def standardise_unit(unit):
    unit = u.Unit(unit)
    bases, powers = [], []
    for base, power in zip(unit.bases, unit.powers):
        if str(base) not in {'ph', 'ct'}:
            bases.append(base)
            powers.append(power)

    return u.CompositeUnit(scale=unit.scale, bases=bases, powers=powers)
```
Is this now good, or is there still a better way to write it?

---

@mhvk - Concerning the suggestion to use `QTable`, again I was tempted. Actually we had one catalog class using QTable and ~ 5 others using Table here, so I played with QTable a bit.
In the end, for now, I decided to drop QTable and use Table instead.

The problem is that Gammapy is heavily based on Table, and we use Table in ~ 100 places.
So I feel that we should either go all-in on Table, or all-in on QTable.

I don't understand why QTable would be better for us. Consider what we'd get if we use QTable for one of the catalogs here:
```
>>> from astropy.table import Table, QTable
>>> QTable.read('https://github.com/gammapy/gammapy-extra/raw/master/datasets/catalogs/fermi/gll_psc_v16.fit.gz').info()
Downloading https://github.com/gammapy/gammapy-extra/raw/master/datasets/catalogs/fermi/gll_psc_v16.fit.gz
|============================================================================================================================================| 2.4M/2.4M (100.00%)         3s
WARNING: hdu= was not specified but multiple tables are present, reading in first available table (hdu=1) [astropy.io.fits.connect]
WARNING: UnitsWarning: 'photon/cm**2/MeV/s' contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: 'photon/cm**2/s' contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
WARNING: UnitsWarning: The unit 'erg' has been deprecated in the FITS standard. Suggested: cm2 g s-2. [astropy.units.format.utils]
WARNING: UnitsWarning: 'erg/cm**2/s' contains multiple slashes, which is discouraged by the FITS standard [astropy.units.format.generic]
<QTable length=3034>
        name          dtype   shape        unit        class   n_bad
-------------------- ------- ------- ---------------- -------- -----
         Source_Name   str18                            Column     0
             RAJ2000 float32                      deg Quantity     0
             DEJ2000 float32                      deg Quantity     0
                GLON float32                      deg Quantity     0
                GLAT float32                      deg Quantity     0
   Conf_68_SemiMajor float32                      deg Quantity    28
   Conf_68_SemiMinor float32                      deg Quantity    28
      Conf_68_PosAng float32                      deg Quantity    28
   Conf_95_SemiMajor float32                      deg Quantity    28
   Conf_95_SemiMinor float32                      deg Quantity    28
      Conf_95_PosAng float32                      deg Quantity    28
             ROI_num   int16                            Column     0
          Signif_Avg float32                            Column     5
        Pivot_Energy float32                      MeV Quantity     0
        Flux_Density float32         ph / (cm2 MeV s) Quantity     0
    Unc_Flux_Density float32         ph / (cm2 MeV s) Quantity     5
            Flux1000 float32             ph / (cm2 s) Quantity     0
        Unc_Flux1000 float32             ph / (cm2 s) Quantity     5
      Energy_Flux100 float32            erg / (cm2 s) Quantity     0
  Unc_Energy_Flux100 float32            erg / (cm2 s) Quantity     5
        Signif_Curve float32                            Column     0
        SpectrumType   str16                            Column     0
      Spectral_Index float32                            Column     0
  Unc_Spectral_Index float32                            Column     9
                beta float32                            Column  2639
            Unc_beta float32                            Column  2680
              Cutoff float32                      MeV Quantity  2918
          Unc_Cutoff float32                      MeV Quantity  2918
           Exp_Index float32                            Column  2918
       Unc_Exp_Index float32                            Column  3028
      PowerLaw_Index float32                            Column     0
          Flux30_100 float32             ph / (cm2 s) Quantity  3034
      Unc_Flux30_100 float32             ph / (cm2 s) Quantity  3034
         nuFnu30_100 float32            erg / (cm2 s) Quantity  3034
       Sqrt_TS30_100 float32                            Column  3034
         Flux100_300 float32             ph / (cm2 s) Quantity     0
     Unc_Flux100_300 float32    (2,)     ph / (cm2 s) Quantity   836
        nuFnu100_300 float32            erg / (cm2 s) Quantity     0
      Sqrt_TS100_300 float32                            Column     5
        Flux300_1000 float32             ph / (cm2 s) Quantity     0
    Unc_Flux300_1000 float32    (2,)     ph / (cm2 s) Quantity   327
       nuFnu300_1000 float32            erg / (cm2 s) Quantity     0
     Sqrt_TS300_1000 float32                            Column     5
       Flux1000_3000 float32             ph / (cm2 s) Quantity     0
   Unc_Flux1000_3000 float32    (2,)     ph / (cm2 s) Quantity   111
      nuFnu1000_3000 float32            erg / (cm2 s) Quantity     0
    Sqrt_TS1000_3000 float32                            Column     0
      Flux3000_10000 float32             ph / (cm2 s) Quantity     0
  Unc_Flux3000_10000 float32    (2,)     ph / (cm2 s) Quantity   137
     nuFnu3000_10000 float32            erg / (cm2 s) Quantity     0
   Sqrt_TS3000_10000 float32                            Column     3
    Flux10000_100000 float32             ph / (cm2 s) Quantity     0
Unc_Flux10000_100000 float32    (2,)     ph / (cm2 s) Quantity   670
   nuFnu10000_100000 float32            erg / (cm2 s) Quantity     0
 Sqrt_TS10000_100000 float32                            Column     2
   Variability_Index float32                            Column     0
         Signif_Peak float32                            Column  2387
           Flux_Peak float32             ph / (cm2 s) Quantity  2387
       Unc_Flux_Peak float32             ph / (cm2 s) Quantity  2387
           Time_Peak float64                        s Quantity  2387
       Peak_Interval float32                        s Quantity  2387
        Flux_History float32   (48,)     ph / (cm2 s) Quantity     0
    Unc_Flux_History float32 (48, 2)     ph / (cm2 s) Quantity 64623
Extended_Source_Name   str18                            Column     0
           0FGL_Name   str17                            Column     0
           1FGL_Name   str18                            Column     0
           2FGL_Name   str18                            Column     0
           1FHL_Name   str18                            Column     0
          ASSOC_GAM1   str15                            Column     0
          ASSOC_GAM2   str14                            Column     0
          ASSOC_GAM3   str15                            Column     0
         TEVCAT_FLAG    str1                            Column     0
           ASSOC_TEV   str21                            Column     0
              CLASS1    str5                            Column     0
              ASSOC1   str26                            Column     0
              ASSOC2   str26                            Column     0
               Flags   int16                            Column     0
```
It's a wild mix of Quantity and Column "columns", a bit hard to keep track of, no?
And if I use a regular table, all columns are Column objects, with the same interface and then with `.quantity` on the column object we have instant access to quantitites when we need them.

IMO in either case (Table or QTable) it would be nice to get something more uniform, e.g. for `Column.quantity` to always work, and just use `u.dimensionless_unscaled` for empty unit string.

I feel the discussion in  http://docs.astropy.org/en/latest/table/mixin_columns.html#quantity-and-qtable should mention `Column.quantity`

@mhvk - Do you think any of these two points (make `.quantity` always work, or improve the docs mentioning the `.quantity` property on Columns) are suggesions worth making in the Astropy tracker?

**Comment by mhvk** [comment]
@cdeil - your implementation is now what I would use myself too!

**Comment by mhvk** [comment]
On `QTable` vs `Table` -- `QTable` is better in that you never (should) have to do `.quantity` -- the columns are just quantities themselves (which is also why I'd be more than a little hesitant to add a `.quantity` property -- I don't really like to burden `Quantity` with crutches for an in hindsight somewhat poor design choice in `astropy.table` -- indeed, I've tried hard to ensure all table-related stuff is in `info` (via `QuantityInfo`). 

Anyway, in general, my hope would be that a user simply does not have to worry about whether something is a quantity or a column.  It is really rather unfortunate that, for backward compatibility, we ended up deciding that that we could not break the pattern where in, e.g., `np.sin(column)`, the unit gets ignored (which is why for a regular table one is forced to do `np.sin(column.quantity)` if one wants a column with units of degrees be treated properly).

Now in the present incarnation of `QTable`, only things with actual dimensions get turned into `Quantity`, the logic partially being that without a unit, a `Column` works fine, and it is the right choice for integer and string. In principle, I guess the default could be that for a float with no unit, one makes a dimensionless `Quantity`, but in practice, the hope at least is that this simply does not matter -- you should not have to care about which class it is stored in.

Overall, I would be quite interested in knowing if anything goes wrong if you just switched to `QTable` throughout... Or do you use masked columns? Because that is the one feature that quantity columns still do not support.

**Comment by cdeil** [comment]
@mhvk - Thanks for the suggestion. I'll try switching to QTable in Gammapy some day and then come back with comments or questions in #980.

**Comment by mhvk** [comment]
I realise this won't be quite trivial, so thanks in advance!

**Pull request merged**